File: apps/api/src/common/errors/app.exception.ts
import { HttpException, HttpStatus } from '@nestjs/common';

import { ERROR_CODES, type ErrorCode } from './error-codes';

export interface AppExceptionBody {
  code: ErrorCode;
  message: string;
  details?: Record<string, string[]>;
}

/**
 * Base class for every deliberate error in the application. Carrying the code
 * on the exception keeps the HTTP contract (`{ code, message }`) consistent
 * without controllers having to shape error responses themselves.
 */
export class AppException extends HttpException {
  readonly code: ErrorCode;
  readonly details?: Record<string, string[]>;

  constructor(
    code: ErrorCode,
    message: string,
    status: HttpStatus,
    details?: Record<string, string[]>,
  ) {
    super({ code, message, details }, status);
    this.code = code;
    this.details = details;
  }
}

export class NotFoundError extends AppException {
  constructor(code: ErrorCode = ERROR_CODES.NOT_FOUND, message = 'Resource not found') {
    super(code, message, HttpStatus.NOT_FOUND);
  }
}

export class ConflictError extends AppException {
  constructor(code: ErrorCode, message: string) {
    super(code, message, HttpStatus.CONFLICT);
  }
}

export class BadRequestError extends AppException {
  constructor(
    code: ErrorCode = ERROR_CODES.BAD_REQUEST,
    message = 'Bad request',
    details?: Record<string, string[]>,
  ) {
    super(code, message, HttpStatus.BAD_REQUEST, details);
  }
}

export class UnauthorizedError extends AppException {
  constructor(code: ErrorCode = ERROR_CODES.UNAUTHORIZED, message = 'Authentication required') {
    super(code, message, HttpStatus.UNAUTHORIZED);
  }
}

export class ForbiddenError extends AppException {
  constructor(code: ErrorCode = ERROR_CODES.FORBIDDEN, message = 'Not allowed') {
    super(code, message, HttpStatus.FORBIDDEN);
  }
}

export class TooManyRequestsError extends AppException {
  constructor(
    code: ErrorCode = ERROR_CODES.RATE_LIMIT_EXCEEDED,
    message = 'Too many requests, please slow down',
  ) {
    super(code, message, HttpStatus.TOO_MANY_REQUESTS);
  }
}

File: apps/api/src/GameScore.Application/Errors/AppException.cs
namespace GameScore.Application.Errors;

public class AppException : Exception
{
    public string Code { get; }
    public int StatusCode { get; }
    public IReadOnlyDictionary<string, string[]>? Details { get; }

    public AppException(string code, string message, int statusCode, IReadOnlyDictionary<string, string[]>? details = null)
        : base(message)
    {
        Code = code;
        StatusCode = statusCode;
        Details = details;
    }
}

public sealed class NotFoundException : AppException
{
    public NotFoundException(string code = ErrorCodes.NotFound, string message = "Resource not found")
        : base(code, message, StatusCodes.Status404NotFound)
    {
    }
}

public sealed class ConflictException : AppException
{
    public ConflictException(string code, string message)
        : base(code, message, StatusCodes.Status409Conflict)
    {
    }
}

public sealed class BadRequestException : AppException
{
    public BadRequestException(
        string code = ErrorCodes.BadRequest,
        string message = "Bad request",
        IReadOnlyDictionary<string, string[]>? details = null)
        : base(code, message, StatusCodes.Status400BadRequest, details)
    {
    }
}

public sealed class UnauthorizedException : AppException
{
    public UnauthorizedException(string code = ErrorCodes.Unauthorized, string message = "Authentication required")
        : base(code, message, StatusCodes.Status401Unauthorized)
    {
    }
}

public sealed class ForbiddenException : AppException
{
    public ForbiddenException(string code = ErrorCodes.Forbidden, string message = "Not allowed")
        : base(code, message, StatusCodes.Status403Forbidden)
    {
    }
}

public sealed class TooManyRequestsException : AppException
{
    public TooManyRequestsException(
        string code = ErrorCodes.RateLimitExceeded,
        string message = "Too many requests, please slow down")
        : base(code, message, StatusCodes.Status429TooManyRequests)
    {
    }
}

file static class StatusCodes
{
    public const int Status400BadRequest = 400;
    public const int Status401Unauthorized = 401;
    public const int Status403Forbidden = 403;
    public const int Status404NotFound = 404;
    public const int Status409Conflict = 409;
    public const int Status429TooManyRequests = 429;
}
